apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransaction.cs

```diff
@@ -401,22 +410,52 @@ Task<IEnumerable<long>> ISpannerTransaction.ExecuteBatchDmlAsync(ExecuteBatchDml
+            // committing/rolling back a transaction.
+            if (_hasCommittedOrRolledBack && _commitResponse != null)
             {
-                var callSettings = SpannerConnection.CreateCallSettings(settings => settings.CommitSettings, CommitTimeout, cancellationToken);
-                var response = await _session.CommitAsync(request, callSettings).ConfigureAwait(false);
-                if (response.CommitTimestamp == null)
-                {
-                    throw new SpannerException(ErrorCode.Internal, "Commit succeeded, but returned a response with no commit timestamp");
-                }
-                if (LogCommitStats)
+                return Task.FromResult(_commitResponse.CommitTimestamp.ToDateTime());
+            }
+            CheckNotHasCommittedOrRolledBack();
+            _hasCommittedOrRolledBack = true;
+            var request = new CommitRequest { Mutations = { _mutations }, ReturnCommitStats = LogCommitStats, RequestOptions = BuildCommitRequestOptions() };
+            return ExecuteHelper.WithErrorTranslationAndProfiling(
+                async () =>
                 {
-                    SpannerConnection.Logger.LogCommitStats(request, response);
-                }
-                return response.CommitTimestamp.ToDateTime();
-            },
-            "SpannerTransaction.Commit", SpannerConnection.Logger);
+                    try
+                    {
+                        var callSettings = SpannerConnection.CreateCallSettings(
+                            settings => settings.CommitSettings, CommitTimeout, cancellationToken);
+                        _commitResponse = await _session.CommitAsync(request, callSettings).ConfigureAwait(false);
+                        if (_commitResponse.CommitTimestamp == null)
+                        {
+                            throw new SpannerException(
+                                ErrorCode.Internal,
+                                "Commit succeeded, but returned a response with no commit timestamp");
+                        }
+
+                        if (LogCommitStats)
+                        {
+                            SpannerConnection.Logger.LogCommitStats(request, _commitResponse);
+                        }
+
+                        return _commitResponse.CommitTimestamp.ToDateTime();
+                    }
+                    finally
+                    {
+                        // Dispose the transaction unless it is a retriable transaction.
+                        // Disposing the transaction will return the session to the pool.
+                        // Retriable transactions reuse the same session for a new transaction if the transaction is aborted
+                        // by Cloud Spanner, and will make sure the session is returned to the pool when done.
+                        if (!_isRetriable)
+                        {
+                            Dispose();
```

Comment:
Agree, but see my comment above. I don't think manually retrying the `Commit`-method is something that we need to support, especially as it would mean that we would not be able to dispose the transaction and return the session to the pool after a valid error message, like `ALREADY_EXISTS` if the user tried to insert a duplicate row using mutations (mutations are sent together with the `Commit` RPC).